README.md
# GWDataFind

The client library for the LIGO Data Replicator (LDR) service.

The DataFind service allows users to query for the location of
Gravitational-Wave Frame (GWF) files containing data from the current
gravitational-wave detectors.

[![PyPI version](https://badge.fury.io/py/gwdatafind.svg)](http://badge.fury.io/py/gwdatafind)
[![Linux status](https://git.ligo.org/lscsoft/gwdatafind/badges/master/pipeline.svg)](https://git.ligo.org/lscsoft/gwdatafind/commits/master)
[![Windows status](https://ci.appveyor.com/api/projects/status/js6gql8960qa9pkl?svg=true)](https://ci.appveyor.com/project/duncanmmacleod/gwdatafind)
[![License](https://img.shields.io/pypi/l/gwdatafind.svg)](https://choosealicense.com/licenses/gpl-3.0/)
[![Documentation status](https://readthedocs.org/projects/gwdatafind/badge/?version=latest)](https://gwdatafind.readthedocs.io/en/latest/?badge=latest)

## Installation

The simplest installation is via `pip`:

```bash
python -m pip install gwdatafind
```

This package is also available as a Conda package:

```bash
conda install -c conda-forge gwdatafind
```

## Basic Usage

To find the URLs of all `H1_R` files for the LIGO-Hanford observatory in
a given GPS interval:

```python
>>> from gwdatafind import connect
>>> conn = connect()
>>> conn.find_urls('H', 'H1_R', 1198800018, 1198800618)
```

This can be shortened for single interactions to

```python
>>> from gwdatafind import find_urls
>>> find_urls('H', 'H1_R', 1198800018, 1198800618)
```

## On the command-line

GWDataFind can also be executed via the command-line client, for full details
run

```bash
$ python -m gwdatafind --help
```

For more documentation, see gwdatafind.readthedocs.org.


debian/changelog
gwdatafind (1.0.4-2) unstable; urgency=low

  * rebuild to separate python library and command-line packages

 -- Duncan Macleod <duncan.macleod@ligo.org>  Fri, 11 Jan 2019 10:14:00 +0000

gwdatafind (1.0.4-1) unstable; urgency=low

  * packaging release to include command-line scripts in python2 package

 -- Duncan Macleod <duncan.macleod@ligo.org>  Fri, 11 Jan 2019 10:14:00 +0000

gwdatafind (1.0.3-1) unstable; urgency=low

  * bug-fix release to support python3 RPMs and testing

 -- Duncan Macleod <duncan.macleod@ligo.org>  Fri, 04 Jan 2019 10:52:51 +0000

gwdatafind (1.0.2-1) unstable; urgency=low

  * bug-fix release

 -- Duncan Macleod <duncan.macleod@ligo.org>  Tue, 14 Aug 2018 12:50:00 +0100

gwdatafind (1.0.1-1) unstable; urgency=low

  * bug-fix release

 -- Duncan Macleod <duncan.macleod@ligo.org>  Tue, 14 Aug 2018 09:52:00 +0100

gwdatafind (1.0.0-1) unstable; urgency=low

  * first release

 -- Duncan Macleod <duncan.macleod@ligo.org>  Mon, 30 Jul 2018 20:45:00 +0100


gwdatafind/tests/test_ui.py
# -*- coding: utf-8 -*-
# Copyright Duncan Macleod 2018
#
# This file is part of GWDataFind.
#
# GWDataFind is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# GWDataFind is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with GWDataFind.  If not, see <http://www.gnu.org/licenses/>.

import os

import pytest

try:
    from unittest import mock
except ImportError:  # python < 3
    import mock

from .. import ui

__author__ = 'Duncan Macleod <duncan.macleod@ligo.org>'

ENVMOCK = None


def setup_module():
    global ENVMOCK
    ENVMOCK = mock.patch.dict(os.environ, {
        'LIGO_DATAFIND_SERVER': 'test.datafind.com',
    })
    ENVMOCK.start()


def teardown_module():
    ENVMOCK.stop()


@mock.patch('gwdatafind.ui.HTTPConnection')
@pytest.mark.parametrize('serv', (
    None,
    'test.datafind.com',
))
def test_connect_http(conn, serv):
    ui.connect(host=serv)
    conn.assert_called_once_with(host='test.datafind.com', port=None)


@mock.patch('ssl.create_default_context')
@mock.patch('gwdatafind.ui.find_credential')
@mock.patch('gwdatafind.ui.HTTPSConnection')
def test_connect_https(conn, cred, sslctx):
    loadcert = mock.MagicMock()
    sslctx.return_value = loadcert
    sslctx
    cred.return_value = ('cert', 'key')

    ui.connect('test.datafind.com:443')

    loadcert.load_cert_chain.assert_called_once_with('cert', 'key')
    conn.assert_called_once_with(
        host='test.datafind.com', port=443,
        context=loadcert,
    )


@mock.patch('gwdatafind.ui.HTTPConnection', return_value=mock.MagicMock())
@pytest.mark.parametrize('method', (
    'ping',
    'find_observatories',
    'find_types',
    'find_times',
    'find_url',
    'find_urls',
    'find_latest',
))
def test_factory_method(conn, method):
    getattr(ui, method)()
    getattr(conn.return_value, method).assert_called_once()




gwdatafind/tests/conftest.py
# -*- coding: utf-8 -*-
# Copyright (C) 2018  Duncan Macleod
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
# Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

"""Test utilities
"""

import pytest

from six.moves import http_client

try:
    from unittest import mock
except ImportError:  # python2.x
    import mock

if pytest.__version__ < '3.0':
    yield_fixture = pytest.yield_fixture
else:
    yield_fixture = pytest.fixture

HTTP_CLIENT = http_client.__name__
HTTP_CONNECTION = '{0}.HTTPConnection'.format(HTTP_CLIENT)


@yield_fixture
def response():
    """Patch an HTTPConnection to do nothing in particular

    Yields the patch for `http.client.HTTPConnection.getresponse`
    """
    with mock.patch('{0}.request'.format(HTTP_CONNECTION)), \
         mock.patch('{0}.getresponse'.format(HTTP_CONNECTION)) as resp:
        yield resp


gwdatafind/tests/test_http.py
# -*- coding: utf-8 -*-
# Copyright (C) 2018  Duncan Macleod
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
# Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

"""Tests for :mod:`gwdatafind.http`
"""

import json
import os
from io import BytesIO
from operator import attrgetter

from six import string_types
from six.moves.urllib.error import HTTPError

try:
    from unittest import mock
except ImportError:
    import mock

import pytest

from ligo.segments import (segment, segmentlist)

from lal.utils import CacheEntry

from .. import utils
from ..http import (
    HTTPConnection,
    HTTPSConnection,
)

LIGO_DATAFIND_SERVER = os.getenv('LIGO_DATAFIND_SERVER')

def setup_module():
    os.environ['LIGO_DATAFIND_SERVER'] = 'test.gwdatafind.com:123'


def teardown_module():
    os.environ.pop('LIGO_DATAFIND_SERVER')
    if LIGO_DATAFIND_SERVER:
        os.environ['LIGO_DATAFIND_SERVER'] = LIGO_DATAFIND_SERVER


def fake_response(output, status=200):
    resp = mock.Mock()
    resp.status = int(status)
    if not isinstance(output, string_types):
        output = json.dumps(output)
    resp.read.return_value = output.encode('utf-8')
    return resp


class TestHTTPConnection(object):
    CONNECTION = HTTPConnection

    @classmethod
    def setup_class(cls):
        cls._create_connection_patch = mock.patch('socket.create_connection')
        cls._create_connection_patch.start()

    @classmethod
    def teardown_class(cls):
        cls._create_connection_patch.stop()

    @classmethod
    @pytest.fixture
    def connection(cls):
        return cls.CONNECTION()

    def test_init(self, connection):
        assert connection.host == 'test.gwdatafind.com'
        assert connection.port == 123

    def test_get_json(self, response, connection):
        response.return_value = fake_response({'test': 1})
        jdata = connection.get_json('something')
        assert jdata['test'] == 1


    def test_ping(self, response, connection):
        response.return_value = fake_response('')
        assert connection.ping() is 0
        response.return_value = fake_response('', 500)
        with pytest.raises(HTTPError):
            connection.ping()

    @pytest.mark.parametrize('match, out', [
        (None, ['A', 'B', 'C', 'D', 'ABCD']),
        ('B', ['B', 'ABCD']),
    ])
    def test_find_observatories(self, response, connection, match, out):
        response.return_value = fake_response(['A', 'B', 'C', 'D', 'ABCD'])
        assert sorted(connection.find_observatories(match=match)) == (
            sorted(out))

    @pytest.mark.parametrize('site, match, out', [
        (None, None, ['A', 'B', 'C', 'D', 'ABCD']),
        ('X', 'B', ['B', 'ABCD']),
    ])
    def test_find_types(self, response, connection, site, match, out):
        response.return_value = fake_response(['A', 'B', 'C', 'D', 'ABCD'])
        assert sorted(connection.find_types(match=match)) == (
            sorted(out))

    def test_find_times(self, response, connection):
        segs = [(0, 1), (1, 2), (3, 4)]
        response.return_value = fake_response(segs)
        times = connection.find_times('X', 'test')
        assert isinstance(times, segmentlist)
        assert isinstance(times[0], segment)
        assert times == segs

        # check keywords
        times = connection.find_times('X', 'test', 0, 10)
        assert times == segs
        with pytest.raises(ValueError):
            times = connection.find_times('X', 'test', gpsstart=0)
        with pytest.raises(ValueError):
            times = connection.find_times('X', 'test', gpsend=10)

    def test_find_url(self, response, connection):
        response.return_value = fake_response(['file:///tmp/X-test-0-10.gwf'])
        url = connection.find_url('X-test-0-10.gwf')
        assert url == [CacheEntry.from_T050017('/tmp/X-test-0-10.gwf')]

        response.return_value = fake_response([])
        with pytest.raises(RuntimeError):
            connection.find_url('X-test-0-10.gwf')
        with pytest.warns(UserWarning):
            url = connection.find_url('X-test-0-10.gwf', on_missing='warn')
            assert url == []
        with pytest.warns(None) as wrngs:
            url = connection.find_url('X-test-0-10.gwf', on_missing='ignore')
            assert url == []
        assert not wrngs.list

    def test_find_frame(self, response, connection):
        response.return_value = fake_response(['file:///tmp/X-test-0-10.gwf'])
        with pytest.warns(DeprecationWarning):
            url = connection.find_frame('X-test-0-10.gwf')
        assert url == [CacheEntry.from_T050017('/tmp/X-test-0-10.gwf')]

    def test_find_latest(self, response, connection):
        response.return_value = fake_response(['file:///tmp/X-test-0-10.gwf'])
        url = connection.find_latest('X', 'test')
        assert url == [CacheEntry.from_T050017('/tmp/X-test-0-10.gwf')]

        response.return_value = fake_response([])
        with pytest.raises(RuntimeError):
            connection.find_latest('X', 'test')
        with pytest.warns(UserWarning):
            url = connection.find_latest('X', 'test', on_missing='warn')
            assert url == []
        with pytest.warns(None) as wrngs:
            url = connection.find_latest('X', 'test', on_missing='ignore')
            assert url == []
        assert not wrngs.list

    def test_find_urls(self, response, connection):
        files =  list(map(CacheEntry.from_T050017, [
            'file:///tmp/X-test-0-10.gwf',
            'file:///tmp/X-test-10-10.gwf',
            'file:///tmp/X-test-20-10.gwf',
        ]))
        response.return_value = fake_response(
            list(map(attrgetter('url'), files)))
        urls = connection.find_urls('X', 'test', 0, 30, match='anything')
        assert urls == files

        # check gaps
        with pytest.raises(RuntimeError):
            connection.find_urls('X', 'test', 0, 40, on_gaps='error')
        with pytest.warns(UserWarning):
            urls = connection.find_urls('X', 'test', 0, 40, on_gaps='warn')
            assert urls == files
        with pytest.warns(None) as wrngs:
            urls = connection.find_urls('X', 'test', 0, 40, on_gaps='ignore')
            assert urls == files
        assert not wrngs.list

    def test_find_frame_urls(self, response, connection):
        files =  list(map(CacheEntry.from_T050017, [
            'file:///tmp/X-test-0-10.gwf',
            'file:///tmp/X-test-10-10.gwf',
            'file:///tmp/X-test-20-10.gwf',
        ]))
        response.return_value = fake_response(
            list(map(attrgetter('url'), files)))
        with pytest.warns(DeprecationWarning):
            urls = connection.find_frame_urls('X', 'test', 0, 30)
        assert urls == files


class TestHTTPSConnection(TestHTTPConnection):
    CONNECTION = HTTPConnection


gwdatafind/ui.py
# -*- coding: utf-8 -*-
# Copyright Duncan Macleod 2018
#
# This file is part of GWDataFind.
#
# GWDataFind is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# GWDataFind is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with GWDataFind.  If not, see <http://www.gnu.org/licenses/>.

import ssl

from .utils import (find_credential, get_default_host)
from .http import (HTTPConnection, HTTPSConnection)

__author__ = 'Duncan Macleod <duncan.macleod@ligo.org>'


def connect(host=None, port=None):
    if host is None:
        host = get_default_host()
    if port is None:
        try:
            host, port = host.rsplit(':', 1)
        except ValueError:
            pass
        else:
            port = int(port)
    if port not in (None, 80):
        cert, key = find_credential()
        context = ssl.create_default_context()
        context.load_cert_chain(cert, key)
        return HTTPSConnection(host=host, port=port, context=context)
    return HTTPConnection(host=host, port=port)


def _with_connection(func):
    def wrapper(*args, **kwargs):
        if not kwargs.get('connection'):
            kw = {key: kwargs.pop(key, None) for key in ('host', 'port')}
            kwargs['connection'] = connect(**kw)
        return func(*args, **kwargs)
    return wrapper


def _ui_factory(target):
    @_with_connection
    def finder(*args, **kwargs):
        connection = kwargs.pop('connection')
        return getattr(connection, target)(*args, **kwargs)

    finder.__doc__ = getattr(HTTPConnection, target).__doc__

    return finder


ping = _ui_factory('ping')
find_observatories = _ui_factory('find_observatories')
find_types = _ui_factory('find_types')
find_times = _ui_factory('find_times')
find_url = _ui_factory('find_url')
find_urls = _ui_factory('find_urls')
find_latest = _ui_factory('find_latest')


gwdatafind/__init__.py
# -*- coding: utf-8 -*-
# Copyright Duncan Macleod 2017
#
# This file is part of GWDataFind.
#
# GWDataFind is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# GWDataFind is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with GWDataFind.  If not, see <http://www.gnu.org/licenses/>.

"""The client library for the LIGO Data Replicator (LDR) service.

The DataFind service allows users to query for the location of
Gravitational-Wave Frame (GWF) files containing data from the current
gravitational-wave detectors.

This package provides the :class:`~HTTPConnection` and
:class:`~HTTPSConnection` class objects, for connecting to an LDR server
in open and authenticated access modes respectively.
The authenticated :class:`~HTTPSConnection` requires users have a valid X509
certificate that is registered with the server in question.

-----------
Quick-start
-----------

The following convenience functions are provided to perform single queries
without a persistent question:


.. currentmodule:: gwdatafind

.. autosummary::
    :nosignatures:

    ping
    find_observatories
    find_types
    find_times
    find_url
    find_urls
    find_latest

For example:

>>> from gwdatafind import find_urls
>>> urls = find_urls("L", "L1_GWOSC_O2_4KHZ_R1", 1187008880, 1187008884,
...                  host="datafind.ligo.org:443")
>>> print(urls)
['file://localhost/cvmfs/gwosc.osgstorage.org/gwdata/O2/strain.4k/frame.v1/L1/1186988032/L-L1_GWOSC_O2_4KHZ_R1-1187008512-4096.gwf']

Additionally, one can manually open a connection using the
:func:`connect` function, and then perform multiple queries.
The :func:`connect` function will automatically select the correct protocol
based on the host given, and will attempt to access any required X509
credentials.

For example:

>>> from gwdatafind import connect
>>> conn = connect(host="datafind.ligo.org", port=443)
>>> obs = conn.find_observatories()
>>> print(obs)
['H', 'V', 'L']
>>> urls = {}
>>> for ifo in obs:
...     urls[ifo] = conn.find_urls(ifo, "{}1_GWOSC_O2_4KHZ_R1".format(ifo),
...                                1187008880, 1187008884)
>>> print(urls)
{'H': ['file://localhost/cvmfs/gwosc.osgstorage.org/gwdata/O2/strain.4k/frame.v1/H1/1186988032/H-H1_GWOSC_O2_4KHZ_R1-1187008512-4096.gwf'],
 'V': ['file://localhost/cvmfs/gwosc.osgstorage.org/gwdata/O2/strain.4k/frame.v1/V1/1186988032/V-V1_GWOSC_O2_4KHZ_R1-1187008512-4096.gwf'],
 'L': ['file://localhost/cvmfs/gwosc.osgstorage.org/gwdata/O2/strain.4k/frame.v1/L1/1186988032/L-L1_GWOSC_O2_4KHZ_R1-1187008512-4096.gwf']}
"""  # noqa: E501

from .http import *
from .ui import *

__author__ = 'Duncan Macleod <duncan.macleod@ligo.org>'
__credits__ = 'Scott Koranda <scott.koranda@ligo.org>'
__version__ = '1.0.4'


gwdatafind/http.py
# -*- coding: utf-8 -*-
# Copyright (C) 2012-2015  Scott Koranda, 2015+ Duncan Macleod
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
# Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

"""Connection utilities for the GW datafind service.
"""

from __future__ import (absolute_import, division)

import os
import re
import socket
import warnings
from json import loads

from six.moves import http_client
from six.moves.urllib.error import HTTPError
from six.moves.urllib.parse import urlparse

from ligo import segments

from .utils import (get_default_host, file_segment)

__author__ = 'Duncan Macleod <duncan.macleod@ligo.org>'
__all__ = ['DEFAULT_SERVICE_PREFIX', 'HTTPConnection', 'HTTPSConnection']

DEFAULT_SERVICE_PREFIX = "/LDR/services/data/v1"


class HTTPConnection(http_client.HTTPConnection):
    """Connect to a GWDataFind host using HTTP.

    Parameters
    ----------
    host : `str`
        the name of the server with which to connect.

    port : `int`, optional
        the port on which to connect.

    **kwargs
        other keywords are passed directly to `http.client.HTTPConnection`
    """
    def __init__(self, host=None, port=None,
                 timeout=socket._GLOBAL_DEFAULT_TIMEOUT, source_address=None):
        """Create a new connection.
        """
        if host is None:
            host = get_default_host()
        http_client.HTTPConnection.__init__(self, host, port,
                                            timeout, source_address)

    def _request_response(self, method, url, **kwargs):
        """Internal method to perform request and verify reponse.

        Parameters
        ----------
        method : `str`
            name of the method to use (e.g. ``'GET'``).

        url : `str`
            remote URL to query.

        **kwargs
            other keyword arguments are passed to
            :meth:`http.client.HTTPConnection.request`.

        Returns
        -------
        response : `str`
            reponse from server query

        Raises
        ------
        RuntimeError
            if query is unsuccessful
        """
        self.request(method, url, **kwargs)
        response = self.getresponse()
        if response.status != 200:
            raise HTTPError(url, response.status, response.reason,
                            response.getheaders(), response.fp)
        return response

    def get_json(self, url, **kwargs):
        """Perform a 'GET' request and return the decode the result as JSON

        Parameters
        ----------
        url : `str`
            remote URL to query

        **kwargs
            other keyword arguments are passed to
            :meth:`HTTPConnection._request_response`

        Returns
        -------
        data : `object`
            JSON decoded using :func:`json.loads`
        """
        response = self._request_response('GET', url, **kwargs).read()
        if isinstance(response, bytes):
            response = response.decode('utf-8')
        return loads(response)

    def get_urls(self, url, scheme=None, on_missing='ignore', **kwargs):
        """Perform a 'GET' request and return a list of URLs.

        Parameters
        ----------
        url : `str`
            remote URL to query

        scheme : `str`, `None`, optional
            the URL scheme to match, default: `None`

        on_missing : `str`, optional
            how to handle an empty (but successful) response, one of

            - ``'ignore'``: do nothing, return empty `list`
            - ``'warn'``: print warning, return empty `list`
            - ``'raise'``: raise `RuntimeError`

        **kwargs
            other keyword arguments are passed to
            :meth:`HTTPConnection.get_json`

        Returns
        -------
        urls : `list` of `str`
            a list of file paths as returned from the server.
        """
        urls = self.get_json(url, **kwargs)

        # sieve for correct file scheme
        if scheme:
            urls = list(filter(lambda e: urlparse(e).scheme == scheme, urls))

        # handle empty result
        if not urls:
            err = "no files found"
            if on_missing == 'warn':
                warnings.warn(err)
            elif on_missing != 'ignore':
                raise RuntimeError(err)

        return urls

    # -- supported interactions -----------------

    def ping(self):
        """Ping the LDR host to test for life.

        Raises
        ------
        RuntimeError
            if the ping fails
        """
        url = '{prefix}/gwf/H/R/1,2'.format(prefix=DEFAULT_SERVICE_PREFIX)
        self._request_response("HEAD", url)
        return 0

    def find_observatories(self, match=None):
        """Query the LDR host for observatories.

        Parameters
        ----------
        match : `str`, `re.Pattern`
            restrict returned observatories to those matching a
            regular expression.

        Returns
        -------
        obs : `list` of `str`
            the list of known osbervatory prefices (and combinations)

        Examples
        --------
        >>> from gwdatafind import connect
        >>> conn = connect()
        >>> conn.find_observatories()
        ['AGHLT', 'G', 'GHLTV', 'GHLV', 'GHT', 'H', 'HL', 'HLT',
         'L', 'T', 'V', 'Z']
        >>> conn.find_observatories("H")
        ['H', 'HL', 'HLT']
        """
        url = "%s/gwf.json" % DEFAULT_SERVICE_PREFIX
        sitelist = set(self.get_json(url))
        if match:
            regmatch = re.compile(match)
            return [site for site in sitelist if regmatch.search(site)]
        return list(sitelist)

    def find_types(self, site=None, match=None):
        """Query the LDR host for frame types.

        Parameters
        ----------
        site : `str`
            single-character name of site to match

        match : `str`, `re.Pattern`
            regular expression to match against known types

        Returns
        -------
        types  : `list` of `str`
            list of frame types

        Examples
        --------
        >>> from gwdatafind import connect
        >>> conn = connect()
        >>> conn.find_types("L", "RDS")
        ['L1_RDS_C01_LX',
         'L1_RDS_C02_LX',
         'L1_RDS_C03_L2',
         'L1_RDS_R_L1',
         'L1_RDS_R_L3',
         'L1_RDS_R_L4',
         'PEM_RDS_A6',
         'RDS_R_L1',
         'RDS_R_L2',
         'RDS_R_L3',
         'TESTPEM_RDS_A6']
        """
        if site:
            url = "%s/gwf/%s.json" % (DEFAULT_SERVICE_PREFIX, site[0])
        else:
            url = "%s/gwf/all.json" % DEFAULT_SERVICE_PREFIX
        typelist = set(self.get_json(url))
        if match:
            regmatch = re.compile(match)
            return [type_ for type_ in typelist if regmatch.search(type_)]
        return list(typelist)

    def find_times(self, site, frametype, gpsstart=None, gpsend=None):
        """Query the LDR for times for which files are avaliable.

        Parameters
        ----------
        site : `str`
            single-character name of site to match

        frametype : `str`
            name of frametype to match

        start : `int`
            GPS start time of query

        end : `int`
            GPS end time of query

        Returns
        -------
        segments : `ligo.segments.segmentlist`
            the list of `[start, stop)` intervals for which files are
            available.
        """
        if gpsstart is not None and gpsend is not None:
            url = ("{prefix}/gwf/{site}/{type}/segments/"
                   "{start},{end}.json".format(
                       prefix=DEFAULT_SERVICE_PREFIX, site=site,
                       type=frametype, start=gpsstart, end=gpsend))
        elif gpsstart is not None or gpsend is not None:
            raise ValueError("please give both `gpsstart` and `gpsend`")
        else:
            url = "{prefix}/gwf/{site}/{type}/segments.json".format(
                prefix=DEFAULT_SERVICE_PREFIX, site=site, type=frametype)

        segmentlist = self.get_json(url)
        return segments.segmentlist(map(segments.segment, segmentlist))

    def find_url(self, framefile, urltype='file', on_missing="error"):
        """Query the LDR host for a single filename.

        Parameters
        ----------
        frametype : `str`
            name of frametype to match

        urltype : `str`, optional
            file scheme to search for, one of ``'file'``, ``'gsiftp'``, or
            `None`, default: 'file'

        on_missing : `str`
            what to do when the requested file isn't found, one of:

            - ``'warn'``: print a warning (default),
            - ``'error'``: raise a `RuntimeError`, or
            - ``'ignore'``: do nothing

        Returns
        -------
        urls : `list` of `str`
            a list of structured file paths for all instances of ``filename``.
        """
        framefile = os.path.basename(framefile)

        # parse file name for site, frame type (expects T050017)
        site, frametype, _, _ = framefile.split("-")

        # query
        url = "{prefix}/gwf/{site}/{type}/{filename}.json".format(
            prefix=DEFAULT_SERVICE_PREFIX, site=site, type=frametype,
            filename=framefile)
        return self.get_urls(url, scheme=urltype, on_missing=on_missing)

    def find_frame(self, *args, **kwargs):
        """DEPRECATED, use :meth:`~HTTPConnection.find_url` instead.
        """
        warnings.warn('find_frame() was renamed find_url()',
                      DeprecationWarning)
        return self.find_url(*args, **kwargs)

    def find_latest(self, site, frametype, urltype='file', on_missing="error"):
        """Query for the most recent file of a given type.

        Parameters
        ----------
        site : `str`
            single-character name of site to match

        frametype : `str`
            name of frametype to match

        urltype : `str`, optional
            file scheme to search for, one of 'file', 'gsiftp', or
            `None`, default: 'file'

        on_missing : `str`, optional
            what to do when the requested frame isn't found, one of:

            - ``'warn'`` print a warning (default), or
            - ``'error'``: raise a `RuntimeError`, or
            - ``'ignore'``: do nothing

        Returns
        -------
        latest : `list` with one `str`
            the URLs of the latest file found (all file types)

        Raises
        ------
        RuntimeError
            if no frames are found
        """
        url = '{prefix}/gwf/{site}/{type}/latest{urltype}.json'.format(
            prefix=DEFAULT_SERVICE_PREFIX, site=site, type=frametype,
            urltype='/{0}'.format(urltype) if urltype else '',
        )
        return self.get_urls(url, scheme=urltype, on_missing=on_missing)

    def find_urls(self, site, frametype, gpsstart, gpsend,
                  match=None, urltype='file', on_gaps="warn"):
        """Find all files of the given type in the [start, end) GPS interval.

        site : `str`
            single-character name of site to match

        frametype : `str`
            name of frametype to match

        gpsstart : `int`
            integer GPS start time of query

        gpsend : `int`
            integer GPS end time of query

        match : `str`, `re.Pattern`, optional
            regular expression to match against

        urltype : `str`, optional
            file scheme to search for, one of 'file', 'gsiftp', or
            `None`, default: 'file'

        on_gaps : `str`, optional
            what to do when the requested frame isn't found, one of:

            - ``'warn'`` print a warning (default), or
            - ``'error'``: raise a `RuntimeError`, or
            - ``'ignore'``: do nothing

        Returns
        -------
        cache : `list` of `str`
            the list of discovered file URLs
        """
        url = '{prefix}/gwf/{site}/{type}/{start},{end}{urltype}.json'.format(
            prefix=DEFAULT_SERVICE_PREFIX, site=site, type=frametype,
            start=gpsstart, end=gpsend,
            urltype='/{0}'.format(urltype) if urltype else '',
        )

        # append a regex if input
        if match:
            url += "?match={0}".format(match)

        # make query
        urls = self.get_urls(url)

        # ignore missing data
        if on_gaps == "ignore":
            return urls

        # handle missing data
        span = segments.segment(gpsstart, gpsend)
        seglist = segments.segmentlist(map(file_segment, urls)).coalesce()
        missing = (segments.segmentlist([span]) - seglist).coalesce()
        if not missing:  # no gaps
            return urls

        # warn or error on missing
        msg = "Missing segments: \n%s" % "\n".join(map(str, missing))
        if on_gaps == "warn":
            warnings.warn(msg)
            return urls
        raise RuntimeError(msg)

    def find_frame_urls(self, *args, **kwargs):
        """DEPRECATED, use :meth:`~HTTPConnection.find_urls` instead.
        """
        warnings.warn('find_frame_urls() was renamed find_urls()',
                      DeprecationWarning)
        return self.find_urls(*args, **kwargs)


class HTTPSConnection(http_client.HTTPSConnection, HTTPConnection):
    """Connect to a GWDataFind host using HTTPS.

    This requires a valid X509 credential registered with the remote host.

    Parameters
    ----------
    host : `str`
        the name of the server with which to connect.

    port : `int`, optional
        the port on which to connect.

    **kwargs
        other keywords are passed directly to `http.client.HTTPSConnection`
    """
    def __init__(self, host=None, port=None, **kwargs):
        """Create a new connection.
        """
        if host is None:
            host = get_default_host()
        http_client.HTTPSConnection.__init__(self, host, port=port, **kwargs)
